drivers/crc/crc_hw.h

```diff
@@ -0,0 +1,253 @@
+#endif /* CONFIG_CRC_THREADSAFE */
+};
+
+#ifdef CONFIG_CRC_HW
+
+typedef int (*crc_api_setup_crc_8_ccitt)(const struct device *dev);
+typedef uint8_t (*crc_api_crc_8_ccitt)(const struct device *dev, uint8_t initial_value,
+				       const void *buf, size_t len);
+typedef uint8_t (*crc_api_verify_crc_8_ccitt)(const struct device *dev, uint8_t expected,
+					      uint8_t initial_value, const void *buf, size_t len);
+typedef int (*crc_api_setup_crc_32_ieee)(const struct device *dev);
+typedef uint32_t (*crc_api_crc_32_ieee)(const struct device *dev, const void *buf, size_t len);
+typedef uint32_t (*crc_api_verify_crc_32_ieee)(const struct device *dev, uint32_t expected,
+					       const void *buf, size_t len);
+
+/**
+ * @struct crc_driver_api
+ * @brief Publicly available API accessible as part of device struct
+ */
+struct crc_driver_api {
+	crc_api_setup_crc_8_ccitt setup_crc_8_ccitt;
+	crc_api_crc_8_ccitt crc_8_ccitt;
+	crc_api_verify_crc_8_ccitt verify_crc_8_ccitt;
+	crc_api_setup_crc_32_ieee setup_crc_32_ieee;
+	crc_api_crc_32_ieee crc_32_ieee;
+	crc_api_verify_crc_32_ieee verify_crc_32_ieee;
+};
+
+#endif /* CONFIG_CRC_HW */
+
+#ifdef CONFIG_CRC_HW
+
+#ifdef CONFIG_CRC_THREADSAFE
+
+/**
+ * @brief Attempt to acquire CRC mutex lock
+ * @param timeout Time in milliseconds to wait, or K_NO_WAIT / K_FOREVER
+ * @retval 0 Mutex successfully acquired
+ * @retval -EBUSY Returned without waiting
+ * @retval -EAGAIN Timeout after waiting for timeout period
+ */
+static inline int crc_lock(k_timeout_t timeout)
+{
+	static const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(crc1));
+	struct crc_data *data = (struct crc_data *)dev->data;
+
+	return k_mutex_lock(&data->mutex, timeout);
+}
+
+/**
+ * @brief Unlock the CRC mutex
+ * @retval 0 Mutex successfully unlocked
+ * @retval -EPERM Current thread does not own mutex
+ * @retval -EINVAL Mutex is not currently locked
+ */
+static inline int crc_unlock(void)
+{
+	static const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(crc1));
+	struct crc_data *data = (struct crc_data *)dev->data;
+
+	struct k_mutex crc_mtx = data->mutex;
+	int res = 0;
+
+	/**
+	 * The lock count can be greater than 1 if the same thread acquires
+	 * a lock multiple times, so we have to iterate until lock_count
+	 * is zero
+	 */
+	while (crc_mtx.lock_count != 0) {
+		res = k_mutex_unlock(&data->mutex);
+
+		if (res != 0) {
+			return res;
+		}
+	}
+
+	return res;
+}
+
+#endif /* CONFIG_CRC_THREADSAFE */
+
+/**
+ * @brief Configure CRC unit to calculate CRC 8 CCITT
+ * @retval See crc_lock()
+ */
+static inline int setup_crc_8_ccitt(void)
+{
+	static const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(crc1));
+	const struct crc_driver_api *api = (const struct crc_driver_api *)dev->api;
+
+	/**
+	 * Attempt to take mutex lock, if we fail then return crc_lock() retvals,
+	 * else return the setup function which will return 0
+	 */
+#ifdef CONFIG_CRC_THREADSAFE
+	const struct crc_data *data = dev->data;
```

Comment:
AFAIK, `device->data` is driver-private and opaque to Zephyr. I don't think using it for this purpose is OK.